File: hamburgeria/flutter_totem/lib/services/socket_service.dart
// lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../config/constants.dart';

class SocketService {
  static final SocketService _instance = SocketService._internal();
  factory SocketService() => _instance;
  SocketService._internal();

  late IO.Socket socket;

  void connect() {
    socket = IO.io(
      AppConfig.socketUrl,
      IO.OptionBuilder()
          .setTransports(['websocket'])
          .disableAutoConnect()
          .build(),
    );
    socket.connect();

    socket.onConnect((_) {
      print('✅ WebSocket connesso');
    });

    socket.onDisconnect((_) {
      print('❌ WebSocket disconnesso');
    });
  }

  void onMenuUpdated(Function(dynamic) callback) {
    socket.on('menu_updated', callback);
  }

  void onOrderUpdated(Function(dynamic) callback) {
    socket.on('order_updated', callback);
  }

  void onOrderNew(Function(dynamic) callback) {
    socket.on('order_new', callback);
  }

  void disconnect() {
    socket.disconnect();
  }
}

File: hamburgeria/angular_staff/src/app/services/socket.service.ts
// src/app/services/socket.service.ts
import { Injectable } from '@angular/core';
import { io, Socket } from 'socket.io-client';
import { Observable, Subject } from 'rxjs';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class SocketService {
  private socket: Socket;
  private orderNew$ = new Subject<any>();
  private orderUpdated$ = new Subject<any>();
  private menuUpdated$ = new Subject<any>();

  constructor() {
    this.socket = io(environment.socketUrl, { transports: ['websocket'] });

    this.socket.on('connect', () =>
      console.log('✅ Staff WebSocket connesso')
    );
    this.socket.on('order_new', (data) => this.orderNew$.next(data));
    this.socket.on('order_updated', (data) => this.orderUpdated$.next(data));
    this.socket.on('menu_updated', (data) => this.menuUpdated$.next(data));
  }

  get onOrderNew(): Observable<any> { return this.orderNew$.asObservable(); }
  get onOrderUpdated(): Observable<any> { return this.orderUpdated$.asObservable(); }
  get onMenuUpdated(): Observable<any> { return this.menuUpdated$.asObservable(); }

  disconnect() { this.socket.disconnect(); }
}
